src/shared/logs/batches.py

```diff
@@ -0,0 +1,143 @@
+from datetime import datetime
+from typing import Literal
+
+from pydantic import BaseModel
+
+from shared.logs.events import (
+    Maintainer,
+    RawEventType,
+    RawMaintainerEvent,
+    RawPackageEvent,
+    RawStatusEvent,
+    sort_events_chronologically,
+)
+
+
+class FoldedEvent(BaseModel):
+    """Base class for folded events that can represent single or bulk operations."""
+
+    suggestion_id: int
+    timestamp: datetime  # Timestamp of the most recent event of the collection
+    username: str
+
+
+class FoldedStatusEvent(FoldedEvent):
+    """A folded status change event (always singular)."""
+
+    action: Literal["insert", "update"]
+    status_value: str
+
+
+class FoldedPackageEvent(FoldedEvent):
+    """A folded package event that can represent single or bulk operations."""
+
+    action: Literal["package.add", "package.remove"]
+    package_names: list[str]  # Always a list, even for single packages
+
+
+class FoldedMaintainerEvent(FoldedEvent):
+    """A folded maintainer event that can represent single or bulk operations."""
+
+    action: Literal["maintainers.add", "maintainers.remove"]
+    maintainers: list[Maintainer]  # Always a list, even for single maintainers
+
+
+FoldedEventType = FoldedStatusEvent | FoldedPackageEvent | FoldedMaintainerEvent
+
+
+def batch_events(
+    events: list[RawEventType], pre_sort: bool = False
+) -> list[FoldedEventType]:
+    """
+    Batch consecutive events of the same type from the same user into bulk operations.
+    Events must be sorted chronologically. Use the pre_sort flag if not.
```

Comment:
What is the use of `pre_sort`? To avoid performing work if the events are already sorted? The wording makes it sound like there are cases where we don't want events to be sorted _at all_, which I guess is not the case. If my initial guess is right, then I would suggest to make it clear, like: "Set the pre_sort flag to False if the events are already sorted", or something like that.

Reply:
I renamed it to `sort`, it makes for sense indeed.

The operations that have a `pre_sort` flag (for now, remove events that cancel each other, and group similar events together) require sorted lists of events, and preserve that order. So when we chain those operations, it isn't useful to re-sort. Hence the flag to be used in operations after the first one.